File: sorts/shell_sort.py
import sys
import time

def main():
    if len(sys.argv) != 2:
        print("Usage: shell_sort.py <input_file>")
        sys.exit(1)

    file_path = sys.argv[1]

    numbers = read_file(file_path)
    start_time = time.time()
    sort(numbers)
    end_time = time.time()
    elapsed_time = (end_time - start_time) * 1000

    print(f"Python Shell sort: {int(elapsed_time)}ms")

def read_file(file_path):
    numbers = []
    with open(file_path, 'r') as file:
        numbers.extend(int(line.strip()) for line in file)
    return numbers

def sort(arr):
    length = len(arr)
    
    # h positions to skip each iteration of h sort
    h_sort = 1
    while h_sort < length / 3:
        h_sort = 3 * h_sort + 1
    
    while h_sort >= 1:
        for i in range(h_sort, length):
            temp = arr[i]
            j = i
            while j >= h_sort and arr[j - h_sort] > temp:
                arr[j] = arr[j - h_sort]
                j -= h_sort
            arr[j] = temp
            
            
        h_sort //= 3 
    
if __name__ == "__main__":
    main()
        
    

File: sorts/ShellSort.java
import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

public class ShellSort implements Comparable<ShellSort> {

    private int value;

    public ShellSort(int value) {
        this.value = value;
    }

    public static void main(String[] args) {
        int length = Integer.parseInt(args[0]);
        ShellSort[] arr = new ShellSort[length];

        int i = 0;
        while(!StdIn.isEmpty()) {
            int num = StdIn.readInt();
            arr[i] = new ShellSort(num);
            i++;
        }
        long startTime = System.nanoTime();
        sort(arr);
        Long endTime = System.nanoTime();
        long elapsedTime = (endTime - startTime) / 1_000_000; // Convert nanoseconds to milliseconds
        StdOut.println("Java Shell sort: " + elapsedTime + "ms");
        // for (ShellSort element : arr) {
        //     StdOut.println(element.value);
        // }
        
    }
    
    public static void sort(Comparable<ShellSort>[] numbers) {
        int N = numbers.length;
        // // Shell sort is insertion sort by h positions
        // // H is determined as non trivial increment value
        for (int h = N / 3; h > 0; h /= 3) {
            for (int i = h; i < N; i++) {
                for (int j = i; j >= h && less(numbers[j], numbers[j - h]); j -= h) {
                    exch(numbers, j , j - h);
                }
            }
        }
        // Shell Sort on while loop
        // VVVVVVVVVVVVVVV
        // VVVVVVVVVVVVVVV
        // VVVVVVVVVVVVVVV
        // int h = 1;
        // while (h < N / 3) {
        //     h = 3 * h + 1;
        // }
        // while (h >= 1) {
        //     // h-sort the array
        //     for (int i = h; i < N; i++) {
        //         for (int j = i; j >= h && less(numbers[j], numbers[j - h]); j -= h) {
        //             exch(numbers, j , j - h);
        //         }
        //     }
        //     h = h / 3;
        // }
    }
    private static void exch(Comparable<ShellSort>[] a, int i, int j) {
        Comparable<ShellSort> swap = a[i];
        a[i] = a[j];
        a[j] = swap;
    }
    private static boolean less(Comparable<ShellSort> v, Comparable<ShellSort> w) {
        return v.compareTo((ShellSort) w) < 0;
    }
    public int compareTo(ShellSort other) {
        return Integer.compare(this.value, other.value);
    }
}